### PR TITLE
Sourcing Environment Variables From Hooks

### LINKED ISSUE #57: [FEATURE]  Inherenting pre_deploy.sh environment variables
## Description

I've just started using dotter, and first off must say I quite like it! Because I'm new, forgive me if I'm missing something.

I arrived at a need to do this as follows: when using dotter to place in `userChrome.css` for Firefox, the profile directory is not always name-static. It's located in `~/.mozilla/firefox/$PROFILE_WHATEVER.default/chrome/userChrome.css`.

The most straightforward solution is to allow some way for scripting the file targets. This most easily can be done with the `pre_deploy.sh` script, which also allows for additional user-configurable erorr checking.

## Implementation Suggestion

After research into previous issues on this repo and doing some light testing, I've come up with a somewhat clean way to do this. **This solution works for unix only**, but I'm pretty sure it can be done in Windows too.

This functionality can be achieved as follows:
1) Instead of directly executing `pre_deploy.sh`, run it through a wrapper
2) The wrapper should call `pre_deploy.sh` in such a way to preserve its environment variables
   - This can be done either with `source` or the `. cmd` syntax in `sh`
   - This should be done through `sh` because dotter already uses it and it's a POSIX shell thing
3) Print out the environment variables in such a way that Rust can interpret and forward them to itself

Below is a proof of concept. While it does not get all the edge cases, I've outlined how to avoid all those that I know may exist:

```rust
use std::process::Command;

fn main() {
    let output = Command::new("sh")
        .args(&[
            "-c",
            "
# this could be pre_deploy.sh
# the . means 'run this as a part of the current shell'
# thus the environment variables will be taken in
. $CARGO_MANIFEST_DIR/src/test.sh 2>&1 >/dev/null
# and just print the environment variables
printenv -0

# note: in a real implementation, instead of using /dev/null:
# 1) Create a pipe FD inside of rust
# 2) Merge the FD number into this string as:
#    printenv -0 >&FDNO
# 3) Read back the pipe via rust
# This would preserve pre_deploy.sh output
",
        ])
        .output()
        .expect("failed to execute process");

    let env_vars = output
        .stdout
        .into_iter()
        .fold(vec![vec![]], |mut acc, c| {
            // Because we used printenv -0, everything is separated by null characters
            // Just need to separate on those
            if c == 0 {
                acc.push(vec![]);
            } else {
                acc.last_mut().unwrap().push(c);
            }
            acc
        })
        .into_iter()
        .map(|v| {
            // Now to separate the names from the values
            let envp = std::str::from_utf8(&v).unwrap();
            // posix compliance states that the seperator between env names and env values is the '=' character
            let i: Vec<_> = envp.splitn(2, "=").collect();
            (
                i[0].to_string(),
                i.get(1).map(|s| s.to_string()).unwrap_or(String::new()),
            )
        })
        .collect::<Vec<_>>();

    // As an example, printing out TESTENV
    // It's feasible to just use std::env::set_var over an iterator
    println!(
        "{:?}",
        env_vars
            .into_iter()
            .find(|s| s.0.starts_with("TESTENV"))
            .unwrap()
    );
}
```

## Offer to Contribute
I'd be happy to flesh out the above solution to capture all the edge cases, and contribute it as a full solution to dotter. @SuperCuber you're in charge here -- if you think this feature is reasonable I'll gladly put the effort in. I don't want to waste time if not.

### PR DESCRIPTION
**Pull request by Shizcow**
Closes #57 

## Description
This PR allows for setting environment variables within `pre_deploy` and `pre_undeploy` files, which will then be taken in for the current process.

## Use Example
Say we have the following snippet within a `global.toml` file:
```toml
[default.files]
my_config_file = "$MAGIC_DIR/my_config_file"
```
Before this PR, one would need to export `$MAGIC_DIR` before running. Now, one can have the following `pre_deploy.sh`:
```sh
export MAGIC_DIR=~/target_magic_dir
```
This will now be shell expanded correctly.
## Implementation
Modifications here are done as follow:
1. Instead of parse and generate the config file in one step, do everything __except__ shell expand
2. Run `pre_deploy`, sourcing the contents
3. Now shell expand file names from the config

The sourcing is roughly done as follows:
1. Create a pipe to talk to the child shell, notated here as `FD`
2. Wrap the `sh` call as `sh -c "source pre_deploy.sh"; printenv -0 >&FD`
3. Inspect the read end of `FD`, parsing the output as environment variable pairs
4. use `std::env::set_var` to forward the child's variables to us

## TODO
This PR will be feature-complete when the following is done:
- [x] Make it work on unix
- [ ] Make it work on windows
- [ ] Write tests

I have yet to find a way to `source` environment variables within windows batch files. And given that I don't run windows, it will be a good bit of time before I were to write and test this on my own. Ideas on this point are welcome.

### CODE REVIEW

**Comment by SuperCuber** [comment]
Whoops, I'm kinda new to the UI of PRs hehe. Not sure how to undo this mark

File: src/hooks.rs [hunk 1]
```diff
@@ -31,22 +34,72 @@ pub(crate) fn run_hook(
...
+        .context("spawn shell")?;
+
+    anyhow::ensure!(
+        child.wait().context("wait for child shell")?.success(),
+        "subshell returned error"
+    );
+
+    {
+        let _drop = pipe_writer;
+    }
```
Comment (SuperCuber) [inline src/hooks.rs:86]:
```suggestion
    drop(pipe_writer);
```

**Comment by SuperCuber** [comment]
Looked through the code a bit, I don't really know much about pipes so I'm not sure if this is the best way to implement this.

This is definitely on the path to getting merged - figuring out windows support (or deciding against it) is an important point though.

**Comment by Shizcow** [comment]
> I'm not sure if this is the best way to implement this

Out of all the inter-process communication methods, pipes tick quite a few boxes for being useful here:
- Available natively in sh (this is the biggest point)
- Doesn't require compiling a companion executable
- Don't need to worry about the filesystem (what if `/tmp/` is full?)
- POSIX compliance (should work OOTB on macos)
  - Also has a Windows analogue (less code duplication and testing)
- Has somewhat decent Rust support (compared to, say, shared memory)
- Fairly light-weight, doesn't add a ton of slow-compiling dependencies to dotter

Of course, not trying to say that pipes are the be-all-end-all. I would like to at least move away from anonymous pipes and into named pipes; this would resolve the issue of a script using FD3 as a hard-coded value (it's not __too__ uncommon). The `interprocess` crate has some support for this, and it seems to be somewhat system-independent.

I'll do a bit of research on this front and report back / update implementation if I find compelling evidence against pipes. Open to suggestions.

**Comment by Shizcow** [comment]
Sorry for no update on this in a long time... school is sucking up all of my time.

I've been thinking about this a little bit more. Is there any reason not to just have additional shell expansion syntax? I don't see why `name = "~/$(magic_dir_command)/name"` would be an issue. It would require more Windows/Unix considerations and figuring out what characters need escaped, but it would be much easier/cleaner to implement. However, the difference in syntax between batch and sh may cause considerable issues. Maybe require keys that use command substitution to declare an OS? Just food for thought.

**Comment by SuperCuber** [comment]
This is already kinda planned in #61 - if we add full handlebars rendering, then you can use the [command_output helper](https://github.com/SuperCuber/dotter/wiki/Templating-and-Caching#helpers) for this.

Now that I'm thinking about it, #61 would also make the functionality in #52 redundant - since you can just make the path render to an empty string to disable the file.

---

I'm not sure however how this addresses the use case in the pull request though... Maybe I'm misunderstanding something.

**Comment by Shizcow** [comment]
With recent Firefox updates, I no longer need this for the exact use-case that I'm after. While I still think this may be useful, I am no longer interested in putting in work towards the solution.

**Pull request closed without merging**